python/hail/typ.py

```diff
@@ -19,7 +20,7 @@ def __str__(self):
         return self.msg
 
 
-class Type(object):
+class Type(HistoryMixin):
```

Comment:
This makes me nervous. Doesn't the stringified type for TVariant/Locus/Interval include the genome reference? If so, this is unnecessary.

Reply:
It's true that I could just write the repr for TVariant/Locus... to print out the name of the reference. However, if we do this, then we will not have any information about how the reference was constructed in the history. And the repr for TVariant/Locus would no longer be a valid repr -- i.e. you can't construct a new GenomeReference object with just the name of the reference. If the repr is valid, you'll get something like `GenomeReference.GRCh37()` everywhere or even worse `GenomeReference(name="foo", contigs=["1", "2"...], lengths={"1": 100, "2": 200, ... })` making the history unusable.

My approach was to think of GenomeReference as similar to a HailContext. It's history should be set once on init and it's varid is set to the name of the reference. Here's an example:

```
# 2017-11-14T22:38:15.134288
# version: devel-b35882e

from hail import *

foo = (GenomeReference.from_file(
    file='src/test/resources/fake_ref_genome.json'))

(KeyTable.parallelize(rows=[{'v': Variant.parse(string='1:32:A:T', reference_genome=foo)}], schema=TStruct(names=['v'], types=[TVariant(reference_genome=foo, required=True)]))
 .export(output='/tmp/foo3.tsv'))
```

I could remove the mixin from singleton types, but Array, Dict, etc. would need it and I didn't see the harm in putting the mixin on Type.